caas/kubernetes/provider/application/constraints.go

```diff
@@ -266,35 +281,131 @@ func processPodAffinity(pod *core.PodSpec, affinityLabels map[string]string) err
 	return nil
 }
 
-func configureConstraint(pod *core.PodSpec, resourceName core.ResourceName, value string) (err error) {
+func configureCharmConstraint(pod *core.PodSpec, resourceName core.ResourceName, requestValue, limitValue string) (err error) {
 	if len(pod.Containers) == 0 {
 		return nil
 	}
-	pod.Containers[0].Resources.Requests, err = MergeConstraint(resourceName, value, pod.Containers[0].Resources.Requests)
+	requestResourceQty, err := resource.ParseQuantity(requestValue)
+	if err != nil {
+		return errors.Annotatef(err, "invalid constraint value %q for %q", requestValue, resourceName)
+	}
+	limitResourceQty, err := resource.ParseQuantity(limitValue)
 	if err != nil {
-		return errors.Annotatef(err, "merging request constraint %s=%s", resourceName, value)
+		return errors.Annotatef(err, "invalid constraint value %q for %q", requestValue, resourceName)
+	}
+	for i, container := range pod.Containers {
+		if container.Name != constants.ApplicationCharmContainer {
+			continue
+		}
```

Comment:
There should always be a charm container if we have containers. But to make the logic easier to follow, we could tweak the logic so that we find the charm container first and then operate on it, eg

```
if len(pod.Containers) == 0 {
	return nil
}
var charmContainer v1.Container
for _, container := range pod.Containers {
	if container.Name = constants.ApplicationCharmContainer {
           charmContainer = container
           break
	}
}
if charmConainer == nil {
    return nil
}
...
```